### PR TITLE
Add minecraft chart to incubator.

### PR DESCRIPTION
**Pull request by gtaylor**
Surprise!

### CODE REVIEW

**Comment by viglesiasce** [comment]
@k8s-bot ok to test

File: incubator/minecraft/README.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+- Kubernetes 1.4+ with Beta APIs enabled
+- PV provisioner support in the underlying infrastructure
+
+## Installing the Chart
+
+To install the chart with the release name `my-release`, read the [Minecraft EULA](https://account.mojang.com/documents/minecraft_eula) run:
+
+```bash
+$ helm install --name my-release \
+    --set MinecraftServer.EULA=true incubator/minecraft
```
Comment (prydonius) [inline incubator/minecraft/README.md]: Can you check this works with Helm beta.2 given the following change:
https://github.com/kubernetes/helm/pull/1451

Given this will be interpreted as boolean, and the default is "FALSE" as a string.

Comment (gtaylor) [reply incubator/minecraft/README.md]: @prydonius I did this with the assumption that the `quote` filter here would stringify it:

https://github.com/kubernetes/charts/pull/154/files#diff-c198687fd5a49f25def79b8ce2933bdeR29

It seems to accept "TRUE" or "true" on the server side. Does the quote filter act like I think it does?

Comment (prydonius) [reply incubator/minecraft/README.md]: Just tested and it does indeed quote so it should be good, thanks!

**Comment by prydonius** [comment]
Any chance in a NOTES.txt? Some short next steps for connecting to the Minecraft server (e.g. getting the IP address) and a warning if the EULA is set to false would be great.

**Comment by gtaylor** [comment]
Sure. I just realized that NOTES.txt was a thing. That looks like a good idea.

File: incubator/minecraft/templates/datadir-pvc.yaml [hunk 2]
```diff
@@ -0,0 +1,14 @@
+{{- if .Values.Persistence.Datadir.Enabled -}}
+kind: PersistentVolumeClaim
+apiVersion: v1
+metadata:
+  name: {{ template "fullname" . }}-datadir
+  annotations:
+    volume.alpha.kubernetes.io/storage-class: {{ .Values.Persistence.StorageClass | quote }}
```
Comment (chrislovecnm) [inline incubator/minecraft/templates/datadir-pvc.yaml]: Do you always need a PVC? How would I install this on minikube?

File: incubator/minecraft/templates/deployment.yaml [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+      containers:
+      - name: {{ template "fullname" . }}
+        image: "{{ .Values.Image }}:{{ .Values.ImageTag }}"
+        imagePullPolicy: Always
+        resources:
+          requests:
+            cpu: "{{.Values.Cpu}}"
+            memory: "{{.Values.Memory}}"
+          limits:
+            cpu: "{{.Values.Cpu}}"
```
Comment (chrislovecnm) [inline incubator/minecraft/templates/deployment.yaml]: Is there a reason to tune limits and requests the same?

File: incubator/minecraft/templates/deployment.yaml [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+            memory: "{{.Values.Memory}}"
+        env:
+        - name: EULA
+          value: {{ .Values.MinecraftServer.EULA | quote }}
```
Comment (chrislovecnm) [inline incubator/minecraft/templates/deployment.yaml]: Holycrap I think you get a cookie for the most ENV vars on this repo. Thank goodness for charts.

File: incubator/minecraft/templates/secrets.yaml [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+metadata:
+  name: {{ template "fullname" . }}
+  labels:
+    app: {{ template "fullname" . }}
+    chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"
+    release: "{{ .Release.Name }}"
+    heritage: "{{ .Release.Service }}"
+type: Opaque
+data:
+  rcon-password: {{ default "" .Values.MinecraftServer.Rcon.Password | b64enc | quote }}
```
Comment (chrislovecnm) [inline incubator/minecraft/templates/secrets.yaml]: So secrets are not too secret. Known K8s challenge. Would a configmap be better?

File: incubator/minecraft/values.yaml [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+  # Check accounts against Minecraft account service.
+  OnlineMode: true
+  # If you adjust this, you may need to adjust Memory above to match.
+  JVMOpts: "-Xmx756M -Xms756M"
+  ServiceType: LoadBalancer
+
+  Rcon:
+    # If you enable this, make SURE to change your Password below.
+    Enabled: false
+    Port: 25575
```
Comment (chrislovecnm) [inline incubator/minecraft/values.yaml]: Should we comment out the params that are off by default? Just a thought

**Review by chrislovecnm — CHANGES_REQUESTED** [review]
Thanks for the PR!!! Got some tweaks dos yah

File: incubator/minecraft/templates/datadir-pvc.yaml (hunk 2 above)
Comment (gtaylor) [reply incubator/minecraft/templates/datadir-pvc.yaml]: The only scenario I can think of where you wouldn't want persistence is in testing. You can disable the PVC and use an emptyDir by setting `Values.Persistence.Datadir.Enabled` to false in the event that you don't.

File: incubator/minecraft/templates/secrets.yaml (hunk 5 above)
Comment (gtaylor) [reply incubator/minecraft/templates/secrets.yaml]: They're not too secret, but I'm not sure why I'd make it a configmap instead. It looks like Secrets are used in some of the other charts.

Hopefully the Secret situation improves as Kubernetes matures. That'd be what I'm banking on in using Secrets, as flimsy as they are now.

File: incubator/minecraft/templates/deployment.yaml (hunk 3 above)
Comment (gtaylor) [reply incubator/minecraft/templates/deployment.yaml]: I may be thinking about this wrong, but Minecraft tends to use what you give it. Aggressively. In this case, we've capped it at 756MB, so it's going to take what we give it as the process runs. I left some extra headroom in the numbers just in case.

I set the requests and limits to the same value since the actual usage should be very predictably just under a specific, user-configurable amount. It's kind of like Redis in how it'll chew up what you give it.

Open to other ideas if this isn't OK.

File: incubator/minecraft/values.yaml (hunk 6 above)
Comment (gtaylor) [reply incubator/minecraft/values.yaml]: I took some time looking around at various charts trying to find a unifying style, but it seems like there's a lot of variation. This was one of the things I wasn't sure about.

I don't feel strongly either way. I guess I like the explicit values since I don't have to stick `default ""` gotmpl things everywhere, but that may not be a good enough reason.

**Comment by gtaylor** [comment]
@chrislovecnm Thanks for the review. I'm not 100% sure which of the above are just questions/suggestions vs blockers for merging to incubator. Maybe that'll be more apparent after reading my responses.

As a sidenote, do we have any kind of style guide that I'm missing? I feel like I'm being really arbitrary with my capitalization and even some of the value organization. Some charts use a flat structure, others use something more nested like I am. Some use dropCase, others use CamelCase.

File: incubator/minecraft/templates/deployment.yaml (hunk 3 above)
Comment (chrislovecnm) [reply incubator/minecraft/templates/deployment.yaml]: So I would perfer to set both separately. You set mem in redis to the same value, because of the app. You don't want k8s sneaking mem from it... 

But I am not a minecraft guru.

File: incubator/minecraft/templates/datadir-pvc.yaml (hunk 2 above)
Comment (chrislovecnm) [reply incubator/minecraft/templates/datadir-pvc.yaml]: Perfect missed that

File: incubator/minecraft/templates/secrets.yaml (hunk 5 above)
Comment (chrislovecnm) [reply incubator/minecraft/templates/secrets.yaml]: Yah secrets have challenges. Sounds good to me

File: incubator/minecraft/values.yaml (hunk 6 above)
Comment (chrislovecnm) [reply incubator/minecraft/values.yaml]: We need to add to the docs on a style we recommend. Appreciate you pointing that out.

File: incubator/minecraft/templates/deployment.yaml (hunk 3 above)
Comment (gtaylor) [reply incubator/minecraft/templates/deployment.yaml]: It was much the same when I last ran a Minecraft server (IE: I set this memory limit, it's going to use every bit of it like Redis under traffic).

Would it be OK to split these into separate values but leave the memory == limit, defined separately?

CPU request doesn't have to == limit, but Minecraft is similarly CPU hoggy. I could yank the limit and leave the request for a full CPU.

**Comment by chrislovecnm** [comment]
@gtaylor looks great. I would ask that we break up request and resource limits.

We need to write recommended best practices / standards for this darn repo. There is an issue open. You brought up a bunch of great questions would love your feedback.

File: incubator/minecraft/templates/deployment.yaml (hunk 3 above)
Comment (viglesiasce) [reply incubator/minecraft/templates/deployment.yaml]: @gtaylor rather than having individual values for each you can make the entire dict a value as shown here:
https://github.com/kubernetes/charts/blob/master/stable/wordpress/values.yaml#L92

I prefer that method. Your defaults can stay the same as you have them now.

File: incubator/minecraft/templates/secrets.yaml (hunk 5 above)
Comment (viglesiasce) [reply incubator/minecraft/templates/secrets.yaml]: I agree with @gtaylor that although the implementation is not perfect we should not be banking on a feature that is not ever intended for keeping secrets. We hope that K8s will make this better over time, best to attach ourselves to the proper feature.

File: incubator/minecraft/values.yaml (hunk 6 above)
Comment (viglesiasce) [reply incubator/minecraft/values.yaml]: Again agree with @gtaylor. I prefer an explicit value in the values.yml rather than implicit defaults at each point in the templates.

**Comment by viglesiasce** [comment]
For best practices please send PRs here:
https://github.com/kubernetes/charts/blob/master/CONTRIBUTING.md#technical-requirements

If it is not yet listed there we shouldnt be blocking PRs from making it in.

File: incubator/minecraft/templates/deployment.yaml (hunk 3 above)
Comment (chrislovecnm) [reply incubator/minecraft/templates/deployment.yaml]: Can we add that to the tech standards?

File: incubator/minecraft/templates/secrets.yaml (hunk 5 above)
Comment (chrislovecnm) [reply incubator/minecraft/templates/secrets.yaml]: Indeed

**Comment by chrislovecnm** [comment]
@viglesiasce agreed

**Comment by gtaylor** [comment]
Some re-jiggering in fixup commits above. I've lowercased all of the values, since that's going to look better for using the nifty `toYaml` thing in the deployment (k8s manifests are lowercase). I've also added a NOTES.txt and adjusted the starting memory to 512 MB. This is workable for 1-3 players.

I can rebase+squash these prior to merge. Or a committer can, whatever is best.

**Comment by chrislovecnm** [comment]
Not sure if I have perms here - @k8s-bot ok to test

Not sure what this repo prefers, but you squashing to chunks of work is always awesome.

**Comment by chrislovecnm** [comment]
@k8s-bot ok to test

**Comment by gtaylor** [comment]
Squashed. Now that we've gone through a few rounds of review, should I still be sending this to incubator? Or should I re-adjust this to be in stable? I think it meets all of the technical requirements now.

**Comment by prydonius** [comment]
TLDR; lgtm!

My only concern in moving this to stable is the EULA requirement breaks the following requirement:

> Must successfully launch with default values (helm install .)
> All pods go to the running state
> All services have at least one endpoint

from [CONTRIBUTING.md](https://github.com/kubernetes/charts/blob/master/CONTRIBUTING.md).

However, given the NOTES.txt appropriately warns about this, and it's reasonable to require a EULA, I'm okay with this going into stable.

Just wanted to point it out, since this could cause issues when the CI does more functional testing of Charts.

File: incubator/minecraft/templates/NOTES.txt [hunk 7]
```diff
@@ -0,0 +1,46 @@
+Getting Started:
```
Comment (prydonius) [inline incubator/minecraft/templates/NOTES.txt]: Can you remove this line? It's redundant since the header helm prints out is `NOTES:`.

**Comment by viglesiasce** [comment]
You could make an if block around the deployment when the EULA is set to false, then you wouldnt deploy it at all and the NOTES could further note that you dont even have it deployed yet.

**Comment by prydonius** [comment]
> You could make an if block around the deployment when the EULA is set to false, then you wouldnt deploy it at all and the NOTES could further note that you dont even have it deployed yet.

Actually this is an interesting idea. Then, the command to add the deployment could simply be `helm upgrade <release> --set minecraftServer.eula=true`. Actually, this command should also work for the current case too.

File: incubator/minecraft/templates/NOTES.txt [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+{{- if eq (.Values.minecraftServer.eula | quote) "FALSE" }}
```
Comment (prydonius) [inline incubator/minecraft/templates/NOTES.txt]: This didn't work for me, it should just be

```
{{- if eq .Values.minecraftServer.eula "FALSE" }}
```

Comment (prydonius) [reply incubator/minecraft/templates/NOTES.txt]: Ah the issue is this doesn't work when you use `--set` because it doesn't work on booleans.

Comment (prydonius) [reply incubator/minecraft/templates/NOTES.txt]:
```
{{- if eq (printf "%s" .Values.minecraftServer.eula) "FALSE" }}
```

should work for both cases

**Comment by gtaylor** [comment]
@prydonius I think I got everything addressed in the fixup commits.

**Comment by gtaylor** [comment]
@prydonius I think I got everything addressed in the fixup commits.

File: stable/minecraft/README.md [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+```bash
+$ helm delete my-release
+```
+
+The command removes all the Kubernetes components associated with the chart and deletes the release.
+
+## Configuration
+
+Refer to `values.yaml` for the full run-down on defaults. These are a mixture of Kubernetes and Minecraft-related directives that map to environment variables in the [itzg/minecraft-server](https://hub.docker.com/r/itzg/minecraft-server/) Docker image.
```
Comment (prydonius) [inline stable/minecraft/README.md]: It would be good to add a link to the values.yaml - https://github.com/kubernetes/charts/blob/master/stable/minecraft/values.yaml

File: stable/minecraft/templates/NOTES.txt [hunk 10]
```diff
@@ -1,10 +1,18 @@
...
-####   ERROR: You did not agree to the EULA in your 'helm install' call   ####
-####     helm delete {{ .Release.Namespace }} and re-run install with     ####
-####       --set minecraftServer.eula="true" added (see README.md).       ####
+####   ERROR: You did not agree to the EULA in your 'helm install' call.  ####
 ##############################################################################
-{{- end }}
+
+This deployment will be incomplete and need to be deleted:
+
+    helm delete {{ .Release.Name }}
```
Comment (prydonius) [inline stable/minecraft/templates/NOTES.txt]: LGTM as is, but a quicker way, is to keep the release and add the deployment by performing an upgrade:

```
helm upgrade {{ .Release.Name }} \
  --set minecraftServer.eula=true stable/minecraft
```

works nicely.

Up to you, one potential problem could be that I'm not 100% sure that upgrading a release to the same chart version is actually supported. cc/ @michelleN

**Comment by viglesiasce** [comment]
I've been doing upgrades to the same version over the last few days and havent hit any issues.

**Comment by gtaylor** [comment]
@prydonius Added the link to README.md and updated NOTES.txt as suggested.

**Comment by prydonius** [comment]
@gtaylor lgtm, great work :)!
